### PR TITLE
No distutils.

### PR DESCRIPTION
**Pull request by erick-xanadu**
**Context:**

> 9.20. [distutils.sysconfig](https://docs.python.org/3.11/distutils/apiref.html#module-distutils.sysconfig) — System configuration information
>
> Deprecated since version 3.10: [distutils.sysconfig](https://docs.python.org/3.11/distutils/apiref.html#module-distutils.sysconfig) has been merged into [sysconfig](https://docs.python.org/3.11/library/sysconfig.html#module-sysconfig).

**Description of the Change:** Use sysconfig instead of distutils.sysconfig if it is not available.

**Benefits:** We can make wheels for python 3.12

### CODE REVIEW

**Comment by erick-xanadu** [comment]
I tested locally and succeeded. I will also add the require build wheels tag for this.

File: setup.py [hunk 1]
```diff
@@ -15,7 +15,11 @@
 import glob
 import platform
 import subprocess
-from distutils import sysconfig
+
+try:
+    from distutils import sysconfig
+except ImportError:
+    import sysconfig
```
Comment (dime10) [inline setup.py:22]: Looks like this package was introduced in Python 3.2, can we just use it always?

Comment (erick-xanadu) [inline setup.py:22]:
```suggestion
import sysconfig
```

Comment (erick-xanadu) [reply setup.py]: https://github.com/PennyLaneAI/catalyst/pull/653/commits/fc46723b98bfcca10f294b01a5efead352a962c1

File: .github/workflows/build-wheel-macos-arm64.yaml [hunk 2]
```diff
@@ -339,6 +339,7 @@ jobs:
 
     - name: Build wheel
       run: |
+        python${{ matrix.python_version }} -m pip install setuptools
```
Comment (dime10) [inline .github/workflows/build-wheel-macos-arm64.yaml:342]: `sysconfig` is in the standard library no, or does this installation have a different purpose?

Comment (erick-xanadu) [reply .github/workflows/build-wheel-macos-arm64.yaml:342]: It has a different purpose. pybind11 will attempt to import distutils if setuptools is not installed. This is necessary for intree_extension that we are importing from pybind11

[See here](https://github.com/pybind/pybind11/blob/master/pybind11/setup_helpers.py#L65-L72)

```python
try:
    from setuptools import Extension as _Extension
    from setuptools.command.build_ext import build_ext as _build_ext
except ImportError:
    from distutils.command.build_ext import (  # type: ignore[assignment]
        build_ext as _build_ext,
    )
    from distutils.extension import Extension as _Extension  # type: ignore[assignment]
```

Comment (dime10) [reply .github/workflows/build-wheel-macos-arm64.yaml:342]: Oh good to know, however I have a few questions:
- We should always try to manage dependencies automatically first, and only do manual installations if necessary. I'm a bit surprised that setuptools isn't available on a system to begin with, but we actually already have it as a package build dependency: https://github.com/PennyLaneAI/catalyst/blob/a3562a485b87ec0adcf9317d14dd1b4c618530d3/pyproject.toml#L31
  Do you have any idea why the build prerequisite in the pyproject.toml is not sufficient?
- Why is this only needed on the arm mac runner?

Either way we can merge this to unblock the release, but we should leave a TODO that this should be removed for the above reasons.

Comment (erick-xanadu) [reply .github/workflows/build-wheel-macos-arm64.yaml:342]: > Why is this only needed on the arm mac runner?

This is not only needed on the arm mac runner. I was just testing.

> We should always try to manage dependencies automatically first, and only do manual installations if necessary. I'm a bit surprised that setuptools isn't available on a system to begin with, but we actually already have it as a package build dependency: 

I agree with you, I was surprised as well. I am not the most familiar with the python build ecosystem myself.

Comment (dime10) [reply .github/workflows/build-wheel-macos-arm64.yaml:343]: Thanks, let's go ahead with this temporary fix then :)

Comment (dime10) [reply .github/workflows/build-wheel-macos-arm64.yaml:343]: Just figured out why `setuptools` is not installed when building the wheel, even though it is listed in the pyproject.toml. 

The pyproject.toml build prerequisites are only used if you use `pip`, which we do when installing the package for development, but not when we create the wheel. Here we invoke the setup script directly `python setup.py bdist_wheel`.

Will update.

**Comment by dime10** [comment]
This PR has been superseded by #667, #669, #670, #671

**Pull request closed without merging**